protocols/kad/src/behaviour.rs

```diff
@@ -286,96 +202,116 @@ where
         };
     }
 
-    /// Inner implementation of the constructors.
-    fn new_inner(local_peer_id: PeerId, records: TRecordStorage) -> Self {
-        let query_config = QueryConfig::default();
-
-        Kademlia {
-            kbuckets: KBucketsTable::new(kbucket::Key::new(local_peer_id), Duration::from_secs(60)),   // TODO: constant
-            protocol_name_override: None,
-            queued_events: SmallVec::new(),
-            active_queries: Default::default(),
-            active_writes: Default::default(),
-            connected_peers: Default::default(),
-            pending_rpcs: SmallVec::with_capacity(query_config.parallelism),
-            next_query_id: QueryId(0),
-            values_providers: FnvHashMap::default(),
-            providing_keys: FnvHashSet::default(),
-            refresh_add_providers: Interval::new_interval(Duration::from_secs(60)).fuse(),     // TODO: constant
-            query_config,
-            add_provider: SmallVec::new(),
-            marker: PhantomData,
-            records,
-        }
-    }
-
     /// Returns an iterator over all peer IDs of nodes currently contained in a bucket
     /// of the Kademlia routing table.
     pub fn kbuckets_entries(&mut self) -> impl Iterator<Item = &PeerId> {
         self.kbuckets.iter().map(|entry| entry.node.key.preimage())
     }
 
-    /// Starts an iterative `FIND_NODE` request.
+    /// Performs a lookup for the closest peers to the given key.
     ///
-    /// This will eventually produce an event containing the nodes of the DHT closest to the
-    /// requested `PeerId`.
-    pub fn find_node(&mut self, peer_id: PeerId) {
-        self.start_query(QueryInfoInner::FindPeer(peer_id));
-    }
-
-    /// Starts an iterative `GET_PROVIDERS` request.
-    pub fn get_providers(&mut self, target: Multihash) {
-        self.start_query(QueryInfoInner::GetProviders { target, pending_results: Vec::new() });
+    /// The result of this operation is delivered in [`KademliaEvent::GetClosestPeersResult`].
+    pub fn get_closest_peers<K>(&mut self, key: K)
+    where
+        K: Into<Multihash> + Clone
+    {
+        let multihash = key.into();
+        let info = QueryInfo::GetClosestPeers { key: multihash.clone() };
+        let target = kbucket::Key::new(multihash);
+        let peers = self.kbuckets.closest_keys(&target);
+        let inner = QueryInner::new(info);
+        self.queries.add_iter_closest(target.clone(), peers, inner);
     }
 
-    /// Starts an iterative `GET_VALUE` request.
+    /// Performs a lookup for a record in the DHT.
```

Comment:
I would explicitly write that this is not instant.
```suggestion
    /// Starts performing a lookup for a record in the DHT.
```

Reply:
Hm, I'm not sure this is an improvement. That kind of documentation reminds me of those on `start_send` methods on futures and streams, where it is always required by the user to call another method (e.g. `poll_complete`) to actually finish the operation. Of course all Kademlia operations are driven by `Kademlia::poll()` but that does typically not need to be called by the user, as it is driven through a Swarm. In other words, `Kademlia::put_record` et al are supposed to _do_ these operation, from start to finish, not just to start them. The fact that the result is delivered asynchronously via `KademliaEvent`s (and thus that the operation is itself asynchronous) is (and needs to be) documented, of course.

Put differently, when I read documentation of the form "starts the process of ..." I think "ok, and what do I have to do to finish it?".